### ISSUE TITLE
Warn in readme about storage-incompatible versions

### ISSUE THREAD

**Issue by blakewest** [OpenZeppelin/openzeppelin-contracts-upgradeable#109]
I upgraded from 3.0 ("contracts-ethereum-package") to 3.4 ("contracts-upgradeable"), and my contracts broke, because I was using them in a base contract, and the storage layout of the OZ contracts changed in between these versions. 

I caught it during testing, so not a huge deal, but still, I was surprised and it ended up taking up some time that could have been avoided. I think anytime the storage layout changes for any of these contracts, that should be marked as a breaking change (eg. to 4.0 in this case)

**💻 Environment**

**📝 Details**

Specifically, it looks like at least ReentrancyGuard changed in between the old and new.
Old: https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/release-3.0/contracts/utils/ReentrancyGuard.sol
New: https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/master/contracts/utils/ReentrancyGuardUpgradeable.sol

There may be others, but this one for sure.

**🔢 Code to reproduce bug**
None necessary. 

Thanks!

**Comment by abcoathup** [comment]
Hi @blakewest,

I’m sorry that you had this issue.  

Thanks so much for reporting it! The project owner will review and triage this issue as soon as they can.

**Comment by frangio** [comment]
@blakewest Sorry about this problem.

It is indeed a breaking change for storage layout. The biggest issue is that there is a missing storage gap in the `Initializable` contract which as a result makes the storage layout always incompatible, although there is a workaround. [Please see this post.](https://forum.openzeppelin.com/t/upgrade-to-contracts-upgradeable-3-3-with-eip712/5200/3?u=frangio)

Users should be using a tool like the [OpenZeppelin Upgrades Plugins](https://github.com/OpenZeppelin/openzeppelin-upgrades) to verify storage layout compatibility. If the plugins don't work for you for some reason please share the feedback.

Where do you think would be the best place to communicate that this upgrade is a breaking change now that it has already been release as 3.4? Where would you have seen the warning?

**Comment by blakewest** [comment]
Hey @frangio ,
Yeah we weren't using the OZ upgrades plugin, because previously it didn't work because it had the library limitations. I had actually raised the issue a while back, and it looks like it [has been fixed](https://github.com/OpenZeppelin/openzeppelin-upgrades/issues/144), so perhaps we can use the plugin now. 

In terms of where I would see it now that it's already been released, I probably would have seen it in the README if you had a big note there. Or next release, perhaps turn it into 4.0?
